src/lib/crypto/kimchi_backend/gadgets/keccak.ml

```diff
@@ -0,0 +1,686 @@
+ * end for
+ * pi:
+ * for x = 0 to 4 do
+ *   for y = 0 to 4 do
+ *     | X |  =  | 0  1 |   | x |
+ *     |   |  =  |      | * |   |
+ *     | Y |  =  | 2  3 |   | y |
+ *     A[X,Y] = a[x,y]
+ *   end for
+ * end for
+ * We use the first index of the state array as the x coordinate and the second index as the y coordinate. 
+ *)
+let pi_rho (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (state : Circuit.Field.t State.matrix) : Circuit.Field.t State.matrix =
+  let state_e = state in
+  let state_b = State.zeros (module Circuit) in
+  (* for all x in {0..4} and y in {0..4}: B[y,2x+3y] = ROT(E[x,y], r[x,y]) *)
+  for x = 0 to keccak_dim - 1 do
+    for y = 0 to keccak_dim - 1 do
+      (* No need to use module since this is always positive *)
+      state_b.(y).(((2 * x) + (3 * y)) mod keccak_dim) <-
+        Bitwise.rot64 (module Circuit) state_e.(x).(y) rot_table.(x).(y) Left
+    done
+  done ;
+  state_b
+
+(* 
+ * Fourth step of the compression function of Keccak for 64-bit words.
+ * F[x,y] = B[x,y] xor ((not B[x+1,y]) and B[x+2,y])
+ * It corresponds to the chi algorithm in the Keccak reference.
+ * for y = 0 to 4 do
+ *   for x = 0 to 4 do
+ *     A[x,y] = a[x,y] xor ((not a[x+1,y]) and a[x+2,y])
+ *   end for
+ * end for   
+ *)
+let chi (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (state : Circuit.Field.t State.matrix) : Circuit.Field.t State.matrix =
+  let state_b = state in
+  let state_f = State.zeros (module Circuit) in
+  (* for all x in {0..4} and y in {0..4}: F[x,y] = B[x,y] xor ((not B[x+1,y]) and B[x+2,y]) *)
+  for x = 0 to keccak_dim - 1 do
+    for y = 0 to keccak_dim - 1 do
+      state_f.(x).(y) <-
+        Bitwise.(
+          bxor64
+            (module Circuit)
+            state_b.(x).(y)
+            (band64
+               (module Circuit)
+               (bnot64_unchecked (module Circuit) state_b.((x + 1) mod 5).(y))
+               state_b.((x + 2) mod 5).(y) ))
+    done
+  done ;
+  (* We can use unchecked NOT because the length of the input is constrained to be
+     64 bits thanks to the fact that it is the output of a previous Xor64 *)
+  state_f
+
+(*
+ * Fifth step of the permutation function of Keccak for 64-bit words.
+ * It takes the word located at the position (0,0) of the state and XORs it with the round constant.
+ *)
+let iota (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (state : Circuit.Field.t State.matrix) (round : int) :
+    Circuit.Field.t State.matrix =
+  let open Circuit in
+  (* Round constants for this round for the iota algorithm *)
+  let rc =
+    exists Field.typ ~compute:(fun () ->
+        Common.field_of_hex (module Circuit) round_consts.(round) )
+  in
+  let state_g = state in
+  state_g.(0).(0) <- Bitwise.(bxor64 (module Circuit) state_g.(0).(0) rc) ;
+  (* Check it is the right round constant *)
+  Field.Assert.equal rc
```

Comment:
I mean you only do this once per round, since there are 24 constants. If you want I can try to make it even before in the circuit, so that multiple blocks don't create more of these. But "again and again each time" is not many times fortunately.

Reply:
Yeah I am checking now. In 2 blocks, doing so means saving 10 rows. Not amazing, but good to have. Thanks for the idea, I will push it like that.